frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -2209,223 +2284,282 @@ resolveFunctionByPoisQuery(Context* context,
-    // does not need to have its corresponding function resolved.
-    if (newTfsForInitializer != sig) {
-      auto resolutionByIdCopy = resolutionById;
-      auto resolvedInit = toOwned(new ResolvedFunction(newTfsForInitializer,
-                                  fn->returnIntent(),
-                                  std::move(resolutionByIdCopy),
-                                  resolvedPoiInfo,
-                                  visitor.returnType));
-      QUERY_STORE_RESULT(resolveFunctionByPoisQuery,
-                         context,
-                         resolvedInit,
-                         newTfsForInitializer,
-                         resolvedPoiInfo.poiFnIdsUsed(),
-                         resolvedPoiInfo.recursiveFnsUsed());
-      auto& saved =
-        resolveFunctionByPoisQueryWrapper(context, newTfsForInitializer,
-                                          resolvedPoiInfo);
-      const ResolvedFunction* resultInit = saved.get();
-      QUERY_STORE_RESULT(resolveFunctionByInfoQuery,
-                         context,
-                         resultInit,
-                         newTfsForInitializer,
-                         poiInfo);
-    }
-
-    // If we resolved an initializer, the result should point to the
-    // final, fully instantiated TFS that was created (if there is
-    // one). In other cases, we just use the input signature.
-    auto finalTfs = newTfsForInitializer ? newTfsForInitializer : sig;
-
-    owned<ResolvedFunction> resolved
-        = toOwned(new ResolvedFunction(finalTfs, fn->returnIntent(),
-                  std::move(resolutionById),
-                  resolvedPoiInfo,
-                  visitor.returnType));
-
-    // Store the result in the query under the POIs used.
-    // If there was already a value for this revision, this
-    // call will not update it. (If it did, that could lead to
-    // memory errors).
-    QUERY_STORE_RESULT(resolveFunctionByPoisQuery,
-                       context,
-                       resolved,
-                       sig,
-                       resolvedPoiInfo.poiFnIdsUsed(),
-                       resolvedPoiInfo.recursiveFnsUsed());
+    auto qt = QualifiedType(QualifiedType::VAR, VoidType::get(context));
+    visitor.returnType = std::move(qt);
+  }
 
-  // On this path we are just resolving a normal function.
-  } else if (fn) {
-    ResolutionResultByPostorderID resolutionById;
-    auto visitor = Resolver::createForFunction(context, fn, poiScope, sig,
-                                               resolutionById);
+  if (fn->body()) {
+    fn->body()->traverse(visitor);
+  } else if (!isInitializer) {
+    CHPL_ASSERT(false && "Should only be called on functions!");
+    return nullptr;
+  }
 
-    if (fn->body()) {
-      fn->body()->traverse(visitor);
-    }
+  // There were outer variables we could not type, so give up.
+  if (isNestedFunction && !caller && !visitor.outerVariables.empty()) {
```

Comment:
It's only a failure condition if `caller` is empty because that means the parent stack frames were not available. If they aren't available there's no way to get a type for the outer variables.

Right now the resolver should always have the `CallerDetails` frames available, but there's not really a way for a user to compute them yet. I have ideas for how they could do that, but I haven't tried it yet.

It is possible for the user (and tests) to call e.g., `typedSignatureInitial` or `resolveFunction` on a nested function, and without this best effort code the tests would fail. To fix those test failures I had to make it so that we'd only give up if the nested function actually used outer variables.